examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+            # as it would require fabric to hold a reference to the model, which we don't want to.
+            raise NotImplementedError("BYOT currently does not support FSDP")
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            optimizer = self.fabric.setup_optimizers(optimizer)
+        else:
+            optimizer, scheduler_cfg = self._parse_optimizers_schedulers(model.configure_optimizers())
+            assert optimizer is not None
+            model, optimizer = self.fabric.setup(model, optimizer)
+
+        # assemble state (current epoch and global step will be added in save)
+        state = {"model": model, "optim": optimizer, "scheduler": scheduler_cfg}
+
+        # load last checkpoint if available
+        latest_checkpoint_path = self.get_latest_checkpoint(self.checkpoint_dir)
+        if latest_checkpoint_path is not None:
+            self.load(state, latest_checkpoint_path)
+
+            # check if we even need to train here
+            if self.max_epochs is not None and self.current_epoch >= self.max_epochs:
+                self.should_stop = True
+
+        # Run validation sanity check
+        if self.val_sanity_check and val_loader is not None and not self.should_stop:
+            self.val_loop(model, val_loader, limit_batches=self.val_sanity_check)
+
+        # TODO: should this be a for loop?
+        while not self.should_stop:
+            self.train_loop(
+                model, optimizer, train_loader, limit_batches=self.limit_train_batches, scheduler_cfg=scheduler_cfg
+            )
+
+            if self.should_validate:
+                self.val_loop(model, val_loader, limit_batches=self.limit_val_batches)
+
+            self.step_scheduler(model, scheduler_cfg, level="epoch", current_value=self.current_epoch)
+
+            self.current_epoch += 1
+
+            # stopping condition on epoch level
+            if self.max_epochs is not None and self.current_epoch >= self.max_epochs:
+                self.should_stop = True
+
+            self.save(state)
+
+    def train_loop(
+        self,
+        model: LightningModule,
+        optimizer: torch.optim.Optimizer,
+        train_loader: DataLoader,
+        limit_batches: Union[int, float] = float("inf"),
+        scheduler_cfg: Optional[Mapping[str, Union[LRScheduler, bool, str, int]]] = None,
+    ):
+        self.fabric.call("on_train_epoch_start")
+        iterable = self.pbar_wrapper(
+            train_loader, total=min(len(train_loader), limit_batches), desc=f"Epoch {self.current_epoch}"
+        )
+
+        for batch_idx, batch in enumerate(iterable):
+            # end epoch if stopping training completely or max batches for this epoch reached
+            if self.should_stop or batch_idx >= limit_batches:
+                self.fabric.call("on_train_epoch_end")
+                return
+
+            self.fabric.call("on_train_batch_start", batch, batch_idx)
```

Comment:
This BYOT example is overfit to having a LM. This is okay. My question is do we want one more that does the same functionality but assuming a `nn.Module`?

Reply:
we could definitely add another one, but that one would then be less "hook-based" (which was a requirement from @lantiga ). Given this requirement, a pure nn.Module Trainer didn't make sense to me.